plugins/rest/src/server_command.cpp

```diff
@@ -0,0 +1,312 @@
+CAF_ALLOW_UNSAFE_MESSAGE_TYPE(
+  std::shared_ptr<vast::plugins::rest::restinio_response>);
+
+namespace vast::plugins::rest {
+
+using router_t = restinio::router::express_router_t<>;
+
+using request_dispatcher_actor = system::typed_actor_fwd<
+  caf::reacts_to<atom::request, restinio_response_ptr, api_endpoint,
+                 system::rest_handler_actor>,
+  caf::reacts_to<atom::internal, atom::request, restinio_response_ptr,
+                 api_endpoint, system::rest_handler_actor>>::unwrap;
+
+namespace {
+
+static restinio::http_method_id_t to_restinio_method(vast::http_method method) {
+  switch (method) {
+    case vast::http_method::get:
+      return restinio::http_method_get();
+    case vast::http_method::post:
+      return restinio::http_method_post();
+  }
+  // Unreachable.
+  return restinio::http_method_get();
+}
+
+struct request_dispatcher_state {
+  request_dispatcher_state() = default;
+
+  rest::server_config server_config;
+  authenticator_actor authenticator;
+};
+
+request_dispatcher_actor::behavior_type request_dispatcher(
+  request_dispatcher_actor::stateful_pointer<request_dispatcher_state> self,
+  server_config config, authenticator_actor authenticator) {
+  self->state.server_config = config;
+  self->state.authenticator = authenticator;
+  return {
+    [self](atom::request, restinio_response_ptr& response,
+           api_endpoint& endpoint, system::rest_handler_actor handler) {
+      // Skip authentication if its not required.
+      if (!self->state.server_config.require_authentication) {
+        self->send(self, atom::internal_v, atom::request_v, std::move(response),
+                   std::move(endpoint), std::move(handler));
+        return;
+      }
+      // Ask the authenticator to validate the passed token.
+      auto token = response->request()->header().try_get_field("X-VAST-Token");
+      if (!token) {
+        response->abort(401, "missing token");
+        return;
+      }
+      self
+        ->request(self->state.authenticator, caf::infinite, atom::validate_v,
+                  *token)
+        .then(
+          [self, response, endpoint = std::move(endpoint),
+           handler](bool valid) mutable {
+            if (valid)
+              self->send(self, atom::internal_v, atom::request_v,
+                         std::move(response), std::move(endpoint),
+                         std::move(handler));
+            else
+              response->abort(401, "invalid token");
+          },
+          [self, response](const caf::error& err) mutable {
+            VAST_WARN("{} received error while authenticating request: {}",
+                      *self, err);
+            response->abort(500, "internal server error");
+          });
+    },
+    [self](atom::internal, atom::request, restinio_response_ptr& response,
+           const api_endpoint& endpoint, system::rest_handler_actor handler) {
+      // TODO: Also consider params in the request body here.
```

Comment:
Do we need to implement this before merging, or is this for a follow-up story?

Reply:
It's a nice-to-have, but all parameters are currently correctly documented as query-parameters so it's certainly not a blocker.